dingo/preprocess.py

```diff
@@ -0,0 +1,248 @@
+        
+        self._initial_reactions =  [ reaction.id for reaction in \
+        self._model.reactions ]
+                    
+        return self._initial_reactions
+    
+    
+    def _reaction_bounds_dictionary(self):
+        """
+        A function used to create a dictionary that maps
+        reactions with their corresponding bounds. It is used to
+        later restore bounds to their wild-type values
+        """
+        
+        self._reaction_bounds_dict = { }
+                
+        for reaction_id in self._initial_reactions:
+            bounds = self._model.reactions.get_by_id(reaction_id).bounds
+            self._reaction_bounds_dict[reaction_id] = bounds
+            
+        return self._reaction_bounds_dict
+
+
+    def _essentials(self):
+        """
+        A function used to find all the essential reactions
+        and append them into a list. Essential reactions are
+        the ones that are required for growth. If removed the 
+        objective function gets zeroed.
+        """
+        
+        self._essential_reactions =  [ reaction.id for reaction in \
+        cobra.flux_analysis.find_essential_reactions(self._model) ]
+                    
+        return self._essential_reactions
+    
+
+    def _zero_flux(self):
+        """
+        A function used to find zero-flux reactions.
+        “Zero-flux” reactions cannot carry a flux while maintaining 
+        at least 90% of the maximum growth rate.
+        These reactions have both a min and a max flux equal to 0,
+        when running a FVA analysis with the fraction of optimum set to 90%
+        """
+        
+        tol = 1e-6
+        
+        fva = cobra.flux_analysis.flux_variability_analysis(self._model, fraction_of_optimum=0.9)
+        zero_flux = fva.loc[ (abs(fva['minimum']) < tol ) & (abs(fva['maximum']) < tol)]
+        self._zero_flux_reactions = zero_flux.index.tolist()
+        
+        return self._zero_flux_reactions
+    
+    
+    def _blocked(self):
+        """
+        A function used to find blocked reactions.
+        "Blocked" reactions that cannot carry a flux in any condition.
+        These reactions can not have any flux other than 0
+        """
+        
+        self._blocked_reactions = cobra.flux_analysis.find_blocked_reactions(self._model, open_exchanges=self._open_exchanges)
+        return self._blocked_reactions
+
+
+    def _metabolically_less_efficient(self):
+        """
+        A function used to find metabolically less efficient reactions.
+        "Metabolically less efficient" require a reduction in growth rate if used
+        These reactions are found when running an FBA and setting the  
+        optimal growth rate as the lower bound of the objective function (in 
+        this case biomass production. After running an FVA with the fraction of optimum
+        set to 0.95, the reactions that have no flux are the metabolically less efficient.
+        """
+            
+        tol = 1e-6
```

Comment:
I am not sure if it would be helpful because if user applies a bigger tolerance value (>1e-6) then it looses the criteria and false-positives reactions will appear (this value is generally used to classify these reactions). If user applies a smaller tolerance value, then the same reactions will appear (at least in the models I tested it). May I create an input parameter and warn the user for the looser check in case he provides a value >1e-6 ?